### PR TITLE
sd-resolve coverity and related fixes

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by yuwata** [comment]
Hmm. With this PR, `systemd-timesyncd` cannot resolve NTP server...
```
May 15 17:28:24 systemd[1]: Starting Network Time Synchronization...
May 15 17:28:24 systemd-timesyncd[21056]: Bus bus-api-timesync: changing state UNSET → OPENING
May 15 17:28:24 systemd-timesyncd[21056]: Bus bus-api-timesync: changing state OPENING → AUTHENTICATING
May 15 17:28:24 systemd-timesyncd[21056]: Added new server ntp.kuins.net.
May 15 17:28:24 systemd-timesyncd[21056]: Added new server 0.fedora.pool.ntp.org.
May 15 17:28:24 systemd-timesyncd[21056]: Added new server 1.fedora.pool.ntp.org.
May 15 17:28:24 systemd-timesyncd[21056]: Added new server 2.fedora.pool.ntp.org.
May 15 17:28:24 systemd-timesyncd[21056]: Added new server 3.fedora.pool.ntp.org.
May 15 17:28:24 systemd-timesyncd[21056]: systemd-timesyncd running as pid 21056
May 15 17:28:24 systemd-timesyncd[21056]: Selected server ntp.kuins.net.
May 15 17:28:24 systemd[1]: Started Network Time Synchronization.
May 15 17:28:24 systemd-timesyncd[21056]: Resolving ntp.kuins.net...
May 15 17:28:24 systemd-timesyncd[21056]: Bus bus-api-timesync: changing state AUTHENTICATING → HELLO
May 15 17:28:24 systemd-timesyncd[21056]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n>
May 15 17:28:24 systemd-timesyncd[21056]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RequestName cookie=2 reply_cookie=0 signature=su error-n>
May 15 17:28:24 systemd-timesyncd[21056]: Assertion 'length == req->length' failed at ../src/libsystemd/sd-resolve/sd-resolve.c:323, function handle_request(). Ignoring.
```

File: src/libsystemd/sd-resolve/sd-resolve.c [hunk 1]
```diff
@@ -975,20 +947,20 @@ _public_ int sd_resolve_getaddrinfo(
...
+        req = (AddrInfoRequest) {
+                .node_len = node ? strlen(node)+1 : 0,
+                .service_len = service ? strlen(service)+1 : 0,
 
-        req.header.id = q->id;
-        req.header.type = REQUEST_ADDRINFO;
-        req.header.length = sizeof(AddrInfoRequest) + req.node_len + req.service_len;
+                .header.id = q->id,
+                .header.type = REQUEST_ADDRINFO,
+                .header.length = sizeof(AddrInfoRequest) + req.node_len + req.service_len,
```
Comment (yuwata) [inline src/libsystemd/sd-resolve/sd-resolve.c]: `req` is not initialized yet.

File: src/libsystemd/sd-resolve/sd-resolve.c [hunk 2]
```diff
@@ -778,12 +764,10 @@ static int handle_response(sd_resolve *resolve, const Packet *packet, size_t len
                 size_t l;
                 struct addrinfo *prev = NULL;
 
-                assert(length >= sizeof(AddrInfoResponse));
-                assert(q->type == REQUEST_ADDRINFO);
+                assert_return(length >= sizeof(NameInfoResponse), -EBADMSG);
+                assert_return(q->type == REQUEST_NAMEINFO, -EBADMSG);
```
Comment (yuwata) [inline src/libsystemd/sd-resolve/sd-resolve.c]: `REQUEST_ADDRINFO` instead of `REQUEST_NAMEINFO`.

**Comment by yuwata** [comment]
By fixing the above two points, now systemd-timesyncd works fine.

File: src/libsystemd/sd-resolve/sd-resolve.c [hunk 3]
```diff
@@ -356,21 +357,21 @@ static int handle_request(int out_fd, const Packet *packet, size_t length) {
                union sockaddr_union sa;
                int ret;
 
-               assert(length >= sizeof(NameInfoRequest));
-               assert(length == sizeof(NameInfoRequest) + ni_req->sockaddr_len);
-               assert(sizeof(sa) >= ni_req->sockaddr_len);
+               assert_return(length >= sizeof(NameInfoRequest), -EBADMSG);
+               assert_return(length == sizeof(NameInfoRequest) + ni_req->sockaddr_len, -EBADMSG);
+               assert_return(ni_req->sockaddr_len < sizeof(sa), -EBADMSG);
```
Comment (yuwata) [inline src/libsystemd/sd-resolve/sd-resolve.c]: Is this intended? Not `<=`?

File: src/libsystemd/sd-resolve/sd-resolve.c (hunk 2 above)
Comment (keszybz) [reply src/libsystemd/sd-resolve/sd-resolve.c]: Yikes.

File: src/libsystemd/sd-resolve/sd-resolve.c (hunk 3 above)
Comment (keszybz) [reply src/libsystemd/sd-resolve/sd-resolve.c]: Yes, it should be `<=`.

**Comment by keszybz** [comment]
I screwed this one up, didn't I? Thank you for the review.

Updated.

File: src/libsystemd/sd-resolve/sd-resolve.c [hunk 4]
```diff
@@ -168,8 +168,10 @@ static void resolve_query_disconnect(sd_resolve_query *q);
 #define RESOLVE_DONT_DESTROY(resolve) \
         _cleanup_(sd_resolve_unrefp) _unused_ sd_resolve *_dont_destroy_##resolve = sd_resolve_ref(resolve)
 
-static int send_died(int out_fd) {
+#define ASSIGN_ERRNO(q, val, error, h_error) \
+        ({ (q)->ret = (val); (q)->_errno = (error); (q)->_h_errno = (h_error); })
```
Comment (poettering) [inline src/libsystemd/sd-resolve/sd-resolve.c:172]: to make this even nicer we could use abs() on the error number

**Pull request merged**